### ISSUE TITLE
Reify an intersection type as the run-time type of the promoted variable's value!

### ISSUE THREAD

**Issue by eernstg** [dart-lang/language#3831]
I know, this is crazy, and surely someone is going to tell us that it is costly in terms of the potential for tree shaking, or something like that. Another likely complaint would be that we don't want `Smi` to be the value of any type arguments, ever. Anyway, let's go! ;-)

An intersection type can be used in a way where it needs to be part of an object:

```dart
List<Object?> f<X>(X x) {
  if (x is int) {
    var result = [x]; // Create a list in the empty context, such that `x` determines the type.
    return result;
  }
  return <Object?>[];
}

void main() {
  print(f<num>(2).runtimeType); // `List<num>`.
}
```

If we evaluate an expression like `[1]` in the empty context then the type argument of the list is taken from the statically known type of the element(s), yielding a `List<int>`.

When `[x]` is evaluated, the static type of `x` is `X & int`, that is, an intersection type. If we follow the normal rules we would infer the list literal as `<X & int>[x]`, using the statically known type of the element or elements to select the actual type argument of the list. However, an intersection type can never be an actual type argument (or any other reified type).

So we need to do _something_ in order to obtain a type which is representable at run time. The approach taken here (and mostly, whenever an intersection type needs to be reified) is to forget about the right hand operand, and just use the left hand operand (which is always a type variable). Hence, `<X>[x]`, in spite of the fact that we just confirmed that `x` has the type `int`.

This explains why this program prints `List<num>`, rather than, say, `List<int>`, or even `List<Object?>`.

We treat intersection types slightly differently when it comes to `await` expressions.

```dart
Future<List<Object?>> f<X>(X x) async {
  if (x is Future<int>) {
    var result = [await x];
    return result;
  }
  return <String>[];
}

void main() async {
  var result = await f<Object?>(Future<int>.value(14));
  print(result.runtimeType); // 'List<int>'!
}
```

In this case we erase the intersection type to the right hand operand rather than the left hand operand: In `f`, `x` is promoted to `X & Future<int>`, and that type [derives the future type](https://github.com/dart-lang/language/blob/8194951697a090bc23f6a84e92044d368a00bf12/specification/dartLangSpec.tex#L11752) `Future<int>`, and hence the static type of `await x` is `int`. So the list will be inferred as `<int>[await x]`.

This shows that we have prior art where an intersection type is erased to the type variable (the left hand operand), and other prior art where it is erased to the promoted bound (the right hand operand).

It would actually be possible to reify the intersection type soundly to the run-time type of the value of the promoted variable. Consider the situation where a local variable `x` has type `X` and is then promoted to type `X & B` based on a type test that confirms that the value has type `B` as well. Let `R` be the run-time type of the value of `x`.

Then we know that `x` has type `X` (by soundness, its value has the declared type) and also type `B` (we just tested that), which means that `R <: X` and `R <: B`.

The claim is now that nothing is known about `X & B` which is not also true for `R`, and hence we can soundly reify `X & B` as `R`.

If we do this then the first example would print `List<int>` rather than `List<num>`. This is because `[x]` is inferred as `<R>[x]` where `R` is the run-time type of the value of `x`, which is `int`. So it prints `List<int>`.

The static type of `[x]` _could_ perhaps be `List<X & int>`, because we could now reify intersection types everywhere, and we could thus also allow them to occur as type arguments, etc. However, for now I'll assume that we keep the static type unchanged as `List<X>`.

With the second example, `x` has static type `X & Future<int>` at `await x`. Again, I'd recommend that we keep the static type unchanged such that the static type of `await x` remains `int`. In this case we do not need a reified value for the type `X & Future<int>`.

An interesting consequence of this proposal is that it allows us to obtain access to the actual run-time type of any given object as the value of a type variable:

```dart
extension<X> on X {
  List<X> get listFor => [];
}

List<X> f<X>(X x) {
  if (x is Object) {
    // `x` has static type `X & Object`.
    var result = [x]; // .. but this is `<R>[x]` where `R` is `int`.
  }
}

void main() {
  Object s = "Hello, world!";
  var list = s.listFor;
  print(list.runtimeType); // 'List<String>`.
  list..add(s)..add(s); // OK, statically and dynamically.
}
```

**Comment by eernstg** [comment]
This is not a good idea after all. Thanks to @johnniwinther for a very good discussion about the implications!

The problem is that the proposed reified value of a type of the form `X & B`, though sound, commits to a run-time behavior which is more constrained than the source code seems to indicate.

```dart
void f<X>(X x1, X x2) {
  if (x1 is int && x2 is int) {
    var list = [x1]; // Creates a "`List<x1.runtimeType>`".
    list.add(x2); // May throw because "`x2 is! x1.runtimeType`".
  }
}
```

This behavior is well-defined, but it is not well justified: The information which is available to the developer is that the static type of `x1` _and_ of `x2` is `X & int`, and this would imply that an operation like `[x1].add(x2)` is safe. With the stronger typing where `[x1]` actually creates a list whose type argument is the run-time type of `x1`, the run-time typing requirement on `x2` cannot reasonably be satisfied.

In other words, this kind of typing is inappropriate because developers can't reasonably be expected to write code which is safe.
